client/hooks/useDatastreamOperation.ts

```diff
@@ -0,0 +1,104 @@
+import { useFetchContext } from "../context/FetchContext";
+import { useEditorContext } from "../context/EditorContext";
+import { deleteObjectDatastreamUrl, downloadObjectDatastreamUrl, postObjectDatastreamUrl } from "../util/routes";
+
+const useDatastreamOperation = () => {
+    const {
+        action: { fetchBlob, fetchText },
+    } = useFetchContext();
+    const {
+        state: { currentPid, activeDatastream, datastreamsCatalog },
+        action: { setSnackbarState, toggleDatastreamModal, getCurrentModelsDatastreams },
+    } = useEditorContext();
+
+    const isAllowedMimeType = (mimeType) => {
+        if (!datastreamsCatalog[activeDatastream]) {
+            return false;
+        }
+        const [type, subtype] = mimeType.split("/");
+        const { allowedType, allowedSubtypes } = datastreamsCatalog[activeDatastream].mimetype;
+        return (
+            (allowedType.includes(type) || allowedType.includes("*")) &&
+            (allowedSubtypes.includes(subtype) || allowedSubtypes.includes("*"))
+        );
+    };
+
+    const uploadFile = async (file) => {
+        try {
+            if (!isAllowedMimeType(file.type)) {
+                throw new Error(`Illegal mime type: ${file.type}`);
+            }
+            const body = new FormData();
+            body.append("stream", activeDatastream);
+            body.append("file", file);
+            const text = await fetchText(postObjectDatastreamUrl(currentPid), {
+                method: "POST",
+                body,
+            });
+            await getCurrentModelsDatastreams();
+            setSnackbarState({
+                open: true,
+                message: text,
+                severity: "success",
+            });
+            toggleDatastreamModal();
+        } catch (err) {
+            setSnackbarState({
+                open: true,
+                message: err.message,
+                severity: "error",
+            });
+            toggleDatastreamModal();
+        }
+    };
+
+    const deleteDatastream = async () => {
+        try {
+            const text = await fetchText(deleteObjectDatastreamUrl(currentPid, activeDatastream), {
+                method: "DELETE",
+            });
+            await getCurrentModelsDatastreams();
+            setSnackbarState({
+                open: true,
+                message: text,
+                severity: "success",
+            });
+        } catch (err) {
+            setSnackbarState({
+                open: true,
+                message: err.message,
+                severity: "error",
+            });
+        }
+        toggleDatastreamModal();
+    };
+
+    const downloadDatastream = async (datastream) => {
+        try {
+            const response = await fetchBlob(downloadObjectDatastreamUrl(currentPid, datastream));
+            const mimeType = response?.headers?.get("Content-Type")?.split("/");
```

Comment:
This needs to be made a little smarter, to handle MIME types like `text/xml; charset=utf8`.

Reply:
See [73236a5](https://github.com/FalveyLibraryTechnology/VuDLPrepJS/pull/118/commits/73236a586e6355e49c35fbdaac70a832125192c9) -- I believe this addresses the problem. If you're happy with this, I believe the PR is ready to merge.